File: internal/record/record.go
package record

import (
	"encoding/binary"
	"errors"
	"fmt"
)

var (
	ErrInvalidEncoding = errors.New("invalid record encoding")
)

type Record struct {
	Key       []byte
	Value     []byte
	Timestamp int64
	IsDeleted bool
}

func NewPut(key, value []byte, timestamp int64) Record {
	return Record{
		Key:       append([]byte(nil), key...),
		Value:     append([]byte(nil), value...),
		Timestamp: timestamp,
		IsDeleted: false,
	}
}

func NewDelete(key []byte, timestamp int64) Record {
	return Record{
		Key:       append([]byte(nil), key...),
		Value:     nil,
		Timestamp: timestamp,
		IsDeleted: true,
	}
}

// Payload layout:
// [keyLen uint32][valueLen uint32][timestamp int64][isDeleted byte][key bytes][value bytes]
func (r Record) Encode() ([]byte, error) {
	keyLen := len(r.Key)
	valueLen := len(r.Value)

	total := 4 + 4 + 8 + 1 + keyLen + valueLen
	out := make([]byte, total)

	binary.BigEndian.PutUint32(out[0:4], uint32(keyLen))
	binary.BigEndian.PutUint32(out[4:8], uint32(valueLen))
	binary.BigEndian.PutUint64(out[8:16], uint64(r.Timestamp))

	if r.IsDeleted {
		out[16] = 1
	} else {
		out[16] = 0
	}

	copy(out[17:17+keyLen], r.Key)
	copy(out[17+keyLen:], r.Value)

	return out, nil
}

func Decode(data []byte) (Record, error) {
	if len(data) < 17 {
		return Record{}, fmt.Errorf("%w: too short", ErrInvalidEncoding)
	}

	keyLen := binary.BigEndian.Uint32(data[0:4])
	valueLen := binary.BigEndian.Uint32(data[4:8])
	timestamp := int64(binary.BigEndian.Uint64(data[8:16]))
	isDeleted := data[16] == 1

	expected := 17 + int(keyLen) + int(valueLen)
	if len(data) != expected {
		return Record{}, fmt.Errorf("%w: expected %d bytes, got %d", ErrInvalidEncoding, expected, len(data))
	}

	key := make([]byte, int(keyLen))
	copy(key, data[17:17+int(keyLen)])

	value := make([]byte, int(valueLen))
	copy(value, data[17+int(keyLen):])

	return Record{
		Key:       key,
		Value:     value,
		Timestamp: timestamp,
		IsDeleted: isDeleted,
	}, nil
}

File: internal/record/record.cpp
#include "record.h"

#include <stdexcept>
#include <cstring>

namespace bigdb::record {

namespace {
void PutUint32(std::vector<unsigned char>& out, uint32_t value) {
    out.push_back(static_cast<unsigned char>((value >> 24) & 0xff));
    out.push_back(static_cast<unsigned char>((value >> 16) & 0xff));
    out.push_back(static_cast<unsigned char>((value >> 8) & 0xff));
    out.push_back(static_cast<unsigned char>(value & 0xff));
}

uint32_t ReadUint32(const std::vector<unsigned char>& data, size_t offset) {
    return ((uint32_t(data[offset]) << 24) |
            (uint32_t(data[offset + 1]) << 16) |
            (uint32_t(data[offset + 2]) << 8) |
            uint32_t(data[offset + 3]));
}

void PutUint64(std::vector<unsigned char>& out, uint64_t value) {
    for (int i = 7; i >= 0; --i) {
        out.push_back(static_cast<unsigned char>((value >> (i * 8)) & 0xff));
    }
}

uint64_t ReadUint64(const std::vector<unsigned char>& data, size_t offset) {
    uint64_t value = 0;
    for (size_t i = 0; i < 8; ++i) {
        value = (value << 8) | data[offset + i];
    }
    return value;
}
}  // namespace

Record NewPut(const std::string& key, const std::string& value, int64_t timestamp) {
    return Record{key, value, timestamp, false};
}

Record NewDelete(const std::string& key, int64_t timestamp) {
    return Record{key, "", timestamp, true};
}

std::vector<unsigned char> Encode(const Record& record) {
    std::vector<unsigned char> out;
    out.reserve(4 + 4 + 8 + 1 + record.key.size() + record.value.size());

    PutUint32(out, static_cast<uint32_t>(record.key.size()));
    PutUint32(out, static_cast<uint32_t>(record.value.size()));
    PutUint64(out, static_cast<uint64_t>(record.timestamp));
    out.push_back(record.isDeleted ? 1 : 0);
    out.insert(out.end(), record.key.begin(), record.key.end());
    out.insert(out.end(), record.value.begin(), record.value.end());
    return out;
}

Record Decode(const std::vector<unsigned char>& data) {
    if (data.size() < 17) {
        throw std::runtime_error("invalid record encoding");
    }

    const uint32_t keyLen = ReadUint32(data, 0);
    const uint32_t valueLen = ReadUint32(data, 4);
    const uint64_t timestamp = ReadUint64(data, 8);
    const bool deleted = data[16] != 0;

    const size_t expected = 17 + keyLen + valueLen;
    if (data.size() != expected) {
        throw std::runtime_error("invalid record encoding length");
    }

    Record rec;
    rec.timestamp = static_cast<int64_t>(timestamp);
    rec.isDeleted = deleted;
    rec.key.assign(data.begin() + 17, data.begin() + 17 + keyLen);
    rec.value.assign(data.begin() + 17 + keyLen, data.end());
    return rec;
}

}  // namespace bigdb::record
